### PR TITLE
Add E2EDiagnosticSetting structure and updateDiagnosticSetting in reg…

### PR DESCRIPTION
**Pull request by michaeljqzq**
…istry

 

# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `master` branch. 

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: service/Microsoft.Azure.Devices/HttpRegistryManager.cs [hunk 1]
```diff
@@ -43,6 +43,7 @@ class HttpRegistryManager : RegistryManager
         const string ConfigurationsRequestUriFormat = "/configurations/?top={0}&{1}";
 
         const string ApplyConfigurationOnDeviceUriFormat = "/devices/{0}/applyConfigurationContent?" + ClientApiVersionHelper.ApiVersionQueryString;
+        const string diagnosticModuleId = "$diagnostic";
```
Comment (erich-wang) [inline service/Microsoft.Azure.Devices/HttpRegistryManager.cs]: capital D

File: service/Microsoft.Azure.Devices/HttpRegistryManager.cs [hunk 2]
```diff
@@ -1186,6 +1187,52 @@ public override Task<Twin> ReplaceTwinAsync(string deviceId, string moduleId, st
...
 #endif
 
+#if ENABLE_MODULES_SDK
+        public override Task<E2EDiagnosticSetting> UpdateDiagnosticModuleSettingAsync(string deviceId, E2EDiagnosticSetting setting, string etag)
+        {
+            return this.UpdateDiagnosticModuleSettingAsync(deviceId, setting, etag, CancellationToken.None);
+        }
+
+        public override async Task<E2EDiagnosticSetting> UpdateDiagnosticModuleSettingAsync(string deviceId, E2EDiagnosticSetting setting, string etag, CancellationToken cancellationToken)
+        {
```
Comment (erich-wang) [inline service/Microsoft.Azure.Devices/HttpRegistryManager.cs:1197]: need to check SamplingPercentage value range

File: service/Microsoft.Azure.Devices/HttpRegistryManager.cs [hunk 3]
```diff
@@ -1186,6 +1187,52 @@ public override Task<Twin> ReplaceTwinAsync(string deviceId, string moduleId, st
...
+
+            var twin = new Twin(deviceId);
+            var settingJson = JsonConvert.SerializeObject(setting);
+            twin.Properties.Desired = new TwinCollection(settingJson);
+            twin.ETag = etag;
+            var twinResult = await UpdateTwinAsync(deviceId, diagnosticModuleId, twin, etag);
+            var updatedDiagnosticSetting = JsonConvert.DeserializeObject<E2EDiagnosticSetting>(twinResult.Properties.Desired.ToJson());
+            return updatedDiagnosticSetting;
+        }
+#endif
```
Comment (erich-wang) [inline service/Microsoft.Azure.Devices/HttpRegistryManager.cs:1235]: add space line

File: service/Microsoft.Azure.Devices/test/Microsoft.Azure.Devices.Api.Test/RegistryManagerTests.cs [hunk 4]
```diff
@@ -938,6 +938,50 @@ private Device PrepareTestDevice(int batteryLevel, string firmwareVersion)
...
+            var restOpMock = new Mock<IHttpClientHelper>();
+            restOpMock.Setup(restOp => restOp.PatchAsync<Twin, Twin>(It.IsAny<Uri>(), It.IsAny<Twin>(), It.IsAny<string>(), It.IsAny<PutOperationType>(), It.IsAny<IDictionary<HttpStatusCode, Func<HttpResponseMessage, Task<Exception>>>>(), It.IsAny<CancellationToken>())).ReturnsAsync(twin);
+
+            var registryManager = new HttpRegistryManager(restOpMock.Object, IotHubName);
+            var returnedSetting = await registryManager.UpdateDiagnosticModuleSettingAsync("abcd", settingToReturn, "*");
+            Assert.AreEqual(settingToReturn.Enabled, returnedSetting.Enabled);
+            Assert.AreEqual(settingToReturn.SamplingPercentage, returnedSetting.SamplingPercentage);
+            restOpMock.VerifyAll();
+        }
+
```
Comment (erich-wang) [inline service/Microsoft.Azure.Devices/test/Microsoft.Azure.Devices.Api.Test/RegistryManagerTests.cs:958]: just leave one space line

File: shared/Microsoft.Azure.Devices.Shared/E2EDiagnosticSetting.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        /// E2EDiagnosticSetting's ETag
+        /// </summary>
+        public string ETag { get; set; }
+
+        /// <summary>
+        /// Gets the E2EDiagnosticSetting as a JSON string
+        /// </summary>
+        /// <param name="formatting">Optional. Formatting for the output JSON string.</param>
+        /// <returns>JSON string</returns>
+        //public string ToJson(Formatting formatting = Formatting.None)
```
Comment (erich-wang) [inline shared/Microsoft.Azure.Devices.Shared/E2EDiagnosticSetting.cs]: remove if not use

File: service/Microsoft.Azure.Devices/HttpRegistryManager.cs (hunk 1 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/HttpRegistryManager.cs]: fixed

File: service/Microsoft.Azure.Devices/HttpRegistryManager.cs (hunk 3 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/HttpRegistryManager.cs:1235]: fixed

File: service/Microsoft.Azure.Devices/test/Microsoft.Azure.Devices.Api.Test/RegistryManagerTests.cs (hunk 4 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/test/Microsoft.Azure.Devices.Api.Test/RegistryManagerTests.cs:958]: fixed

File: shared/Microsoft.Azure.Devices.Shared/E2EDiagnosticSetting.cs (hunk 5 above)
Comment (michaeljqzq) [reply shared/Microsoft.Azure.Devices.Shared/E2EDiagnosticSetting.cs]: fixed

File: service/Microsoft.Azure.Devices/HttpRegistryManager.cs (hunk 2 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/HttpRegistryManager.cs:1197]: added in E2EDiagnosticSetting

**Pull request closed without merging**